doc/designs/quic-design/congestion-control.md

```diff
@@ -0,0 +1,229 @@
+
+ - `payload_length` - the maximum length of payload that can be sent
+   in a single UDP packet (not counting UDP and IP headers).
+ - `smoothed_rtt` - the current round trip time of the connection
+   as computed from the acked packets.
+ - `rtt_variance` - the variance of the `smoothed_rtt` value.
+
+Thread handling
+---------------
+
+The `OSSL_CC_DATA` is created with the `new` function per each
+connection. The expectation is that there is only a single thread
+accessing the `ccdata` which should not be a limitation as the calls
+into the module will be done from the packetizer layer of the QUIC
+connection handling and there the eventual writes from multiple threads
+handling individual streams of the connection have to be synchronized
+already to create the packets.
+
+Congestion event handling
+-------------------------
+
+The congestion state of a connection can change only after some event
+happens - i.e., a packet is considered lost meaning the `on_data_lost()`
+is called, or an ack arrives for a packet causing calls to
+`on_data_acked()` or `on_spurious_congestion_event()` functions.
+The congestion control does not produce any timer events by itself.
+
+Exemptions
+----------
+To facilitate probing and to avoid having to always special-case
+probing packets when considering congestion on sending, the
+`set_exemption()` function allows setting number of packets that
+are allowed to be sent overriding the eventual congestion state.
+
+Appendix A
+----------
+
+Proposed header file with comments explaining the individual
+functions. The API is meant to be internal initially so the method
+accessors to set the individual functions will be added later once
+the API is public. Alternatively this might be also implemented
+as fetchable dispatch API.
+
+```C
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/ssl.h>
+
+typedef struct ossl_cc_method_st OSSL_CC_METHOD;
+/*
+ * An OSSL_CC_DATA is an externally defined opaque pointer holding
+ * the internal data of the congestion control method
+ */
+typedef struct ossl_cc_data_st OSSL_CC_DATA;
+
+/*
+ * Once this becomes public API then we will need functions to create and
+ * free an OSSL_CC_METHOD, as well as functions to get/set the various
+ * function pointers....unless we make it fetchable.
+ */
+struct ossl_cc_method_st {
+    /*
+     * Create a new OSSL_CC_DATA object to handle the congestion control
+     * calculations.
+     *
+     * |settings| are mandatory settings that will cause the
+     * new() call to fail if they are not understood).
+     * |options| are optional settings that will not
+     * cause the new() call to fail if they are not understood.
```

Comment:
Can you get example of such settings?

Can the same values be passed to settings and options?

Reply:
I do not have any concrete examples yet. Basically it would be parameters fine-tuning the algorithm behavior. Potentially they would be loaded from some configuration file (or database). I do not see why you'd pass the same values to settings and options.